### PR TITLE
Fix: Mamba2 `norm_before_gate` usage

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Fixes the default value for norm_before_gate (False) in the default config of mamba2. Additionally, implemented the other variation with norm_before_gate = True. Currently, this only affects continuous training from codestral on the fast path with the fused kernel: https://github.com/huggingface/transformers/blob/20a04497a86dc79adb8a93e31325f25255d317e1/src/transformers/models/mamba2/modeling_mamba2.py#L334-L353 
Discovered in https://github.com/huggingface/transformers/pull/32580#discussion_r1712661960

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@molbap @ArthurZucker

### CODE REVIEW

**Comment by vasqu** [comment]
Tbh, we could also just remove `norm_before_gate` in the config and only allow the `False` path. I don't mind either.

**Review by molbap — APPROVED** [review]
Great, thanks a lot! I'm pro-keeping `norm_before_gate` as it's a valid config option in the original code and people could want to experiment on it. pinging @ArthurZucker for final review :)

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 1]
```diff
@@ -248,7 +254,9 @@ def __init__(self, config: Mamba2Config, layer_idx: int):
         A = torch.arange(1, self.num_heads + 1)
         self.A_log = nn.Parameter(torch.log(A))
         self.A_log._no_weight_decay = True
-        self.norm = MambaRMSNormGated(self.intermediate_size, eps=self.layer_norm_epsilon)
+        self.norm = MambaRMSNormGated(
+            self.intermediate_size, eps=self.layer_norm_epsilon, norm_before_gate=config.norm_before_gate
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/mamba2/modeling_mamba2.py]: It's already been in the Mixer class, I just used it from the passed config. Changed it to use `self` now instead :+1:

**Review by ArthurZucker — COMMENTED** [review]
Hey! Not entirely sure this is super `transfomers` friendly, as a research would just copy paste the modelling and add all the if-else! 
our usualy motivation is to add a new model if there is a good pretrained checkpoint that demonstrated the use of this param! 

If there is a strong demand from the community why not, but in general such changes go against the philosphy! 🤗

**Comment by vasqu** [comment]
Hmm, there are currently two problems in the current implementation though:
1. The `norm_before_gate` flag in the config was incorrectly set to True but the implementation follows the path as if it were False. This is problematic as it does not suggest so and enables incorrect codestral training via the fast path (see the initial description and the code snippet - direct passing of the flag).
2. Now that the flag exists, it would be weird that it only supports one path but suggests otherwise.

If I read it correctly, would you prefer removing this flag in the config and adjusting the code to only follow one path (i.e. `norm_before_gate=False`)?

**Comment by ArthurZucker** [comment]
> If I read it correctly, would you prefer removing this flag in the config and adjusting the code to only follow one path (i.e. norm_before_gate=False)?

yes! 🤗 this would be less confusing IMO!

**Comment by vasqu** [comment]
@ArthurZucker Should be good now :eyes:

**Review by ArthurZucker — APPROVED** [review]
Thanks let's update the PR title

**Comment by vasqu** [comment]
I guess the renaming is too late now 😓

**Comment by ArthurZucker** [comment]
No worries 🤗

**Pull request merged**